GhidraPython/goresym_rename.py

```diff
@@ -0,0 +1,174 @@
+            text = currentProgram.memory.getBlock(".text")
+            return text.start.offset-tmva
+
+    return None
+
+def _func_map_estimator(hints):
+    offs = []
+    # Try to guess with known function names
+    grs_funcs = hints["UserFunctions"] + hints["StdFunctions"]
+    # Match function lists and collect possible offsets
+    for f in fm.getFunctions(True):
+        name = f.name
+        if "FUN_" in name or name == "entry":
+            continue
+        for gf in grs_funcs:
+            if name in gf["FullName"]:
+                offs.append(f.entryPoint.offset-gf["Start"])
+
+    winner, winner_count = None, 0
+    # Count given offsets, most counts is our estimate
+    for o in offs:
+        count = offs.count(o)
+        print("Found offset {} with count {}".format(o, count))
+        if winner is None or winner_count < count:
+            winner, winner_count = o, count
+
+    return winner
+
+def estimate_offset(choice, hints):
+    if choice == CHOICE_EST_KNOWN_FUNCS:
+        estimator = _func_map_estimator
+    elif choice == CHOICE_EST_PCLNTAB:
+        estimator = _pclntab_estimator
+    else:
+        estimator = _entry_addr_estimator
+
+    e = estimator(hints)
+    if e is not None:
+        return e
+
+    return 0    
+
+def ask_offset(default):
+    msg = "Please specify optional memory offset for entrypoints, script estimated {} as most probable".format(default)
+    offset = askString("Optional offset", msg, str(default))
+    return int(offset)
+
+def rename_funcs(items, offset):
+    if not iterable(items):
+        return
+
+    for func in items:
+        try:
+            # ' ' is considered as invalid char
+            # https://github.com/NationalSecurityAgency/ghidra/blob/c19276091f274a9ef0850c904c743f61c850854e/Ghidra/Framework/SoftwareModeling/src/main/java/ghidra/program/model/symbol/SymbolUtilities.java#L104
+            addr, name = func["Start"], func["FullName"].replace(" ", "_")
+            entrypoint = toAddr(addr+offset)
+            f = fm.getFunctionAt(entrypoint)
+            if f is None:
+                print("Creating new func at {}".format(entrypoint))
+                createFunction(entrypoint, name)
+            else:
+                print("Renaming func {} to {}".format(f.name, name))
+                f.setName(name, USER_DEFINED)
+        except Exception as e:
+            print(u"error renaming {} at {}: {}".format(name, hex(addr), e))
+
+def annotate(items, offset):
+    if not iterable(items):
+        return
+
+    for i in items:
+        if i["VA"] != 0:
+            # ' ' is considered as invalid char
+            createLabel(toAddr(i["VA"]+offset), i["Str"].replace(" ", "_"), True)
+
+
+grsfile = askFile("GoReSym output file", "Choose GoReSym output file")
+with open(grsfile.absolutePath, "r") as fp:
+    buf = fp.read()
+hints = json.loads(buf)
+
+choice = askChoice("Offset estimation", "Please choose offset estimation strategy", CHOICES, CHOICE_EST_KNOWN_FUNCS)
```

Comment:
is there any way to try all the strategies and give a hint to the user which will probably work best? as someone who would run this script the first time, i would have no idea which to select and would pick one randomly until it worked.

Reply:
Good thinking and thanks for this idea.

Yes, I have added logic that runs through all estimators and counts the number of performed renames and function creations. The renames are most interesting here because they could be considered as "mapping recognized functions found by Ghidra with functions discovered by GoReSym".

Based on the count of renamed functions, the script suggests a strategy that performed the most renames. The stats with renamed and created count for all estimators are printed to the console as well.